notebooks/02-ingestion/10-updating-data.ipynb

```diff
@@ -0,0 +1,1188 @@
+    "   \n",
+    "- Only include the rows from the existing table that fall into the update timeframe and new rows where the join results in NULL for the `t.\"__time\"` column:\n",
+    "\n",
+    "   ```\n",
+    "   WHERE t.\"__time\" IS NULL OR TIME_IN_INTERVAL(t.\"__time\", '2005-11-03T00:00:00.000Z/P1D')  \n",
+    "   ```\n",
+    "- Since it the operation is covering a whole day use PARTITIONED BY clause uses DAY.\n",
+    "- Set the context parameter to use `joinAlgorithm=sortMerge`.\n",
+    "\n",
+    "Run the following SQL to apply the upsert operation:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "62394515-4baa-4766-b57f-6b071c6ddadd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "REPLACE INTO \"example-flights-updates\" \n",
+    "   OVERWRITE \n",
+    "   WHERE \"__time\" >= TIMESTAMP'2005-11-03 00:00:00' AND \"__time\" < TIMESTAMP'2005-11-04 00:00:00'\n",
+    "SELECT\n",
+    "  COALESCE(t.\"__time\", s.\"__time\") as \"__time\",\n",
+    "  COALESCE(t.\"airline\", s.\"airline\") as \"airline\",\n",
+    "  COALESCE(t.\"flight_number\", s.\"flight_number\") as \"flight_number\",\n",
+    "  CASE WHEN (t.\"__time\" IS NULL OR t.\"__time\" = s.\"__time\") THEN s.\"arrival_time\"       \n",
+    "       ELSE t.\"arrival_time\"                                  \n",
+    "  END AS \"arrival_time\",\n",
+    "  CASE WHEN (t.\"__time\" IS NULL OR t.\"__time\" = s.\"__time\") THEN s.\"tail_number\"\n",
+    "       ELSE t.\"tail_number\"    \n",
+    "  END AS \"tail_number\", \n",
+    "  CASE WHEN (t.\"__time\" IS NULL OR t.\"__time\" = s.\"__time\") THEN s.\"origin\"\n",
+    "       ELSE t.\"origin\"          \n",
+    "  END AS \"origin\", \n",
+    "  CASE WHEN (t.\"__time\" IS NULL OR t.\"__time\" = s.\"__time\") THEN s.\"destination\"\n",
+    "       ELSE t.\"destination\"     \n",
+    "  END AS \"destination\", \n",
+    "  CASE WHEN (t.\"__time\" IS NULL OR t.\"__time\" = s.\"__time\") THEN s.\"departure_delay\"\n",
+    "       ELSE t.\"departure_delay\" \n",
+    "  END AS \"departure_delay\"\n",
+    "FROM \"example-flights-updates\" t \n",
+    "  FULL OUTER JOIN\n",
+    "    \"example-upsert-changeset\" s\n",
+    "  ON t.\"__time\"=s.\"__time\"\n",
+    "  AND t.\"airline\"=s.\"airline\"\n",
+    "  AND t.\"flight_number\"=s.\"flight_number\" \n",
+    "WHERE t.\"__time\" IS NULL OR TIME_IN_INTERVAL(t.\"__time\", '2005-11-03T00:00:00.000Z/P1D') \n",
+    "PARTITIONED BY DAY    -- replacing a whole day so partition by day\n",
+    "'''\n",
+    "req = sql_client.sql_request(sql)\n",
+    "req.add_context(\"sqlJoinAlgorithm\", 'sortMerge')\n",
+    "display.run_task(req)\n",
+    "sql_client.wait_until_ready('example-flights-updates')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "faf24eea-956e-4531-8708-f2e4064757fa",
+   "metadata": {},
+   "source": [
+    "Check the data changes and new rows with the same SQL as before:"
```

Comment:
As this is happening at the row level, I really think you need to show the before and after _as rows_ not using aggregates. It becomes an extra leap to have to understand what the aggregates have done, and why we know that the update worked.

Reply:
I disagree because we would be asking them to visually review the difference in hundreds of rows. That's why I used an aggregate, we see the aggregate effect of the updates because the flight_number range has changed, the total number of flights have changed (new rows) and the SUM("departure_delay") shows the update to the existing rows.

Reply:
Could we meet half way and you add a cell above that shows just a selection of the updated rows?

Reply:
Done. I think that by only showing 20 rows it actually works out really well. I sorted them descending on flight number, so you can see the 9 new rows, the 9 rows that were not touched for that same airline and a couple of rows that show the update... it definitely helps the narrative.